src/threading/cthread.cpp

```diff
@@ -0,0 +1,191 @@
+    task tsk;
+
+    run_task(task_pool&, lock&);
+    ~run_task();
+};
+
+// Own a task in flight
+// lock should be passed locked,
+// and will be unlocked after call
+task_pool::run_task::run_task(task_pool& pool, lock& lck):
+    pool{pool},
+    lck{lck},
+    tsk{}
+{
+    std::swap(tsk, pool._tasks.front());
+    pool._tasks.pop_front();
+    
+    lck.unlock();
+    pool._tasks_available.notify_all();
+}
+
+// Release task
+// Call unlocked, returns unlocked
+task_pool::run_task::~run_task() {
+    lck.lock();
+    tsk.second->in_flight--;
+    
+    lck.unlock();
+    pool._tasks_available.notify_all();
+}
+
+// runs forever until quit is true
+void task_pool::run_tasks_loop() {
+    lock lck = {_tasks_mutex, std::defer_lock};
+    while (true) {
+        lck.lock();
+
+        while (! _quit && _tasks.empty()) {
+            _tasks_available.wait(lck);
+        }
+        if (_quit) {
+            return;
+        }
+    
+        run_task run{*this, lck};
+        run.tsk.first();
+    }
+}
+
+// run until out of tasks for a group
+void task_pool::run_tasks_while(task_group* g) {
+    lock lck{_tasks_mutex, std::defer_lock};
+    while (true) {  
+        lck.lock();
+
+        while (! _quit && g->in_flight && _tasks.empty()) {
+            _tasks_available.wait(lck);
+        }
+        if (_quit || ! g->in_flight) {
+            return;
+        }
+
+        run_task run{*this, lck};
+        run.tsk.first();
+    }
+}
+
+// Create pool and threads
+// new threads are nthreads-1
+task_pool::task_pool(std::size_t nthreads):
+    _tasks_mutex{},
+    _tasks_available{},
+    _tasks{},
+    _threads{}
+{
+    assert(nthreads > 0);
+  
+    // now for the main thread
+    auto tid = std::this_thread::get_id();
+    _thread_ids[tid] = 0;
+  
+    // and go from there
+    for (std::size_t i = 1; i < nthreads; i++) {
+        _threads.emplace_back([this]{run_tasks_loop();});
+        tid = _threads.back().get_id();
+        _thread_ids[tid] = i;
+    }
+}
+
+task_pool::~task_pool() {
+    {
+        lock lck{_tasks_mutex};
+        _quit = true;
+    }
+    _tasks_available.notify_all();
+    
+    for (auto& thread: _threads) {
+        thread.join();
+    }
+}
+
+// push a task into pool
+void task_pool::run(const task& tsk) {
+    {
+        lock lck{_tasks_mutex};
+        _tasks.push_back(tsk);
+        tsk.second->in_flight++;
+    }
+    _tasks_available.notify_all();
+}
+
+void task_pool::run(task&& tsk) {
+  {
+      lock lck{_tasks_mutex};
+      _tasks.push_back(std::move(tsk));
+      tsk.second->in_flight++;
+  }
+  _tasks_available.notify_all();
+}
+
+// call on main thread
+// uses this thread to run tasks
+// and waits until the entire task
+// queue is cleared
+void task_pool::wait(task_group* g) {
+    run_tasks_while(g);
+}
+
+class NMC_NUM_THREADS_ERROR {};
+static auto NMC_NUM_THREADS_VAR{"NMC_NUM_THREADS_VAR"};
+static auto NMC_NUM_THREADS{"NMC_NUM_THREADS"};
+static auto OMP_NUM_THREADS{"OMP_NUM_THREADS"};
+
```

Comment:
I admit, I find this more confusing than just using the C string constants directly below in the calls to `std::getenv`.